ambari-common/src/main/python/ambari_commons/repo_manager/command_repository.py

```diff
@@ -0,0 +1,262 @@
+    self.version_id = _find_value(json_dict, 'repoVersionId')
+    self.stack_name = _find_value(json_dict, 'stackName')
+    self.resolved = _find_value(json_dict, 'resolved', False)
+    self.__version_string = _find_value(json_dict, 'repoVersion')
+    self.__stack_tool_version = None  # used as cache for resolved via stack tool version
+    self.repo_filename = _find_value(json_dict, 'repoFileName')
+    self.feat = CommandRepositoryFeature(_find_value(json_dict, "feature", default={}))
+    self.__items = []
+
+    repos_def = _find_value(json_dict, 'repositories')
+    if repos_def is not None:
+      if not isinstance(repos_def, list):
+        repos_def = [repos_def]
+
+      for repo_def in repos_def:
+        self.__items.append(CommandRepositoryItem(self, repo_def))
+
+  @property
+  def version_string(self):
+    """
+    Returning repository version. If repository is not yet resolved, trying to extract version vis stack tool.
+
+    WARNING!!! To be resolved correctly via stack tool, call the property after package installation.
+    """
+    if not self.resolved:
+      if not self.__stack_tool_version:
+        from resource_management.libraries.functions.stack_select import get_role_component_current_stack_version
+        from resource_management.core.exceptions import Fail
+        from resource_management.core.logger import Logger
+
+        #  here is a hack to make all nested logging in "get_role_component_current_stack_version" silenced
+        log_disabled = Logger.is_disabled()  # required to not override top-level log disable call
+        if not log_disabled:
+          Logger.disable()
+
+        try:
+          self.__stack_tool_version = get_role_component_current_stack_version()
+        except Fail:
+          pass
+        finally:
+          if not log_disabled:
+            Logger.enable()
+
+      if self.__stack_tool_version:
+        return self.__stack_tool_version
+
+    return self.__version_string
+
+  @property
+  def all_items(self):
+    """
+    Return all parsed repo entities passed to the command
+
+    :return all repos from the command definition
+    """
+    return self.__items
+
+  @property
+  def items(self):
+    """
+    Returns filtered list of the repositories.
+
+    For the rules check #filter_items function
+
+    :return filtered list of the repositories
+    """
+    return self.filter_items()
+
+  def filter_items(self, exclude_by_tags=()):
+    """
+    Filter items by various conditions
+
+    :type exclude_by_tags set(str)
+    :rtype list[CommandRepositoryItem]
+    """
+    exclude_by_tags = set(exclude_by_tags)
+
+    if self.__service_name_filter:
+      assert isinstance(self.__service_name_filter, (str, unicode))
+
+    exclude_by_tags = exclude_by_tags.union(self.__tag_filter)
+
+    filtered_items = []
+
+    for item in self.__items:
+      exclude_item = False
+      if len(item.tags - (item.tags ^ exclude_by_tags)) > 0:
```

Comment:
what's the difference between 

```python
item.tags - (item.tags ^ exclude_by_tags)
```
and this?
```python
item.tags & exclude_by_tags
```

Reply:
amount of operations